### PR TITLE
Handle exceptions thrown from RestCompatibleVersionHelper

### LINKED ISSUE #78214: Using incorrect Accept/Content-Type headers in compatibility mode has bad client-side UX
Discovered this behavior when testing the compatibility mimetype (`application/vnd.elasticsearch+json?compatible-with=X`) being used with `Accept` and `Content-Type`. When using the compatibility mimetype in either `Accept` or `Content-Type` the other HTTP header if given must be set to the compatibility mimetype as well otherwise an error occurs.

When running the following command:

```
$ curl -k -XPOST -H "Accept: application/vnd.elasticsearch+json;compatible-with=7" \
  -H "Content-Type: application/json" -d '{"query":{"match_all":{}}}' \
  http://localhost:9200/_search
```

The error message server-side in the logs is a good one, can tell exactly what's wrong:

```
A compatible version is required on both Content-Type and Accept headers if either one has requested a compatible version and the compatible versions must match. Accept=application/vnd.elasticsearch+json;compatible-with=7, Content-Type=application/json
```

but client-side this is what's returned:

```
curl: (52) Empty reply from server
```

Elasticsearch doesn't send an HTTP response and instead hangs up the socket. This isn't a great user-experience, perhaps this can be changed to Elasticsearch returning a 4XX HTTP response (maybe 400, 406, 415?) with the error message above?

### LINKED ISSUE #79060: Invalid mediatype with compatible-with throws exception
when an invalid media type is used on content-type or Accept header but with compatible-with parameter an exception is thrown which is not handled and emits a WARN log message
nothing is returned to a server

reproduce:
```
curl --request PUT \
  --url 'http://localhost:9200/my_index1?include_type_name=true' \
  --header 'Accept: aaa/bbb' \
  --header 'Authorization: Basic ZWxhc3RpYzpwYXNzd29yZA==' \
  --header 'Content-Type: aaa/bbb;compatible-with=7' \
  --data '{
    "mappings": {
        "type1": {
            "properties": {
                "location": {
                    "type": "geo_shape"
                }
            }
        }
    }
}'
```

we should handle this exception and return something meaning full to a user

I have a feeling that it might have been discussed already.  @jakelandis @joegallo do you remember? any thoughts?

### PR DESCRIPTION
**Pull request by pgomulka**
RestCompatibleVersionHelper is used to validate versions on Accept and Content-Type headers
When a validation fails, an exception is being thrown indicating which headers are incorrect.
That exception was not handled in RestRequest where the helper was used.
This commit gracefully handles the exception from validation failure. A bad request response is returned to a user.

closes #79060
closes #78214

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/rest/RestCompatibleVersionHelperTests.java [hunk 1]
```diff
@@ -24,6 +24,16 @@
...
+    // public void testCompatibleOnUnkownMediaType() {
+    // RestRequest.MediaTypeHeaderException e = expectThrows(
+    // RestRequest.MediaTypeHeaderException.class,
+    // () -> requestWith(acceptHeader("wrong/type;compatible-with=7"), contentTypeHeader(null), bodyPresent())
+    // );
+    // assertThat(
+    // e.getMessage(),
+    // equalTo("Unknown media type Accept=wrong/type;compatible-with=")
+    // );
+    // }
```
Comment (joegallo) [inline server/src/test/java/org/elasticsearch/rest/RestCompatibleVersionHelperTests.java:36]: This part right here seems off. Perhaps accidentally committed? WIP that was meant to be removed?

File: server/src/main/java/org/elasticsearch/rest/RestRequest.java [hunk 2]
```diff
@@ -607,20 +613,29 @@ public RestApiVersion getRestApiVersion() {
...
+            this.message = message;
+            this.failedHeaderNames = failedHeaderNames;
+        }
 
-        MediaTypeHeaderException(final IllegalArgumentException cause, String failedHeaderName) {
+        MediaTypeHeaderException(final RuntimeException cause, String... failedHeaderNames) {
             super(cause);
-            this.failedHeaderName = failedHeaderName;
+            this.failedHeaderNames = Set.of(failedHeaderNames);
+            this.message = "Invalid media-type value on header " + this.failedHeaderNames;
```
Comment (joegallo) [inline server/src/main/java/org/elasticsearch/rest/RestRequest.java:629]: Little wording/formatting nit -- maybe "on headers [HeaderA, HeaderB, HeaderC]"? What do you think?

**Review by joegallo — CHANGES_REQUESTED** [review]

File: server/src/test/java/org/elasticsearch/rest/RestCompatibleVersionHelperTests.java (hunk 1 above)
Comment (pgomulka) [reply server/src/test/java/org/elasticsearch/rest/RestCompatibleVersionHelperTests.java:36]: ah yes, initially I thought I can include validation for unknown media types before parsing versions, but this is far beyond the scope of this PR.

**Review by jakelandis — APPROVED** [review]
LGTM

**Pull request merged**